reactivesocket-core/src/main/java/io/reactivesocket/internal/PublisherFunctions.java

```diff
@@ -0,0 +1,316 @@
+package io.reactivesocket.internal;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+/**
+ * A set of utility functions for applying function composition over {@link Publisher}s.
+ */
+public final class PublisherFunctions {
+
+    private PublisherFunctions() {
+        // No instances.
+    }
+
+    /**
+     * Converts a {@code Publisher} of type {@code T} to a {@code Publisher} of type {@code R} using the passed
+     * {@code map} function.
+     *
+     * @param source {@code Publisher} to map.
+     * @param map {@code Function} to use for conversion.
+     *
+     * @param <T> Type of source {@code Publisher}.
+     * @param <R> Type of resulting {@code Publisher}.
+     *
+     * @return A new {@code Publisher} which takes objects of type {@code R} instead of {@code T}.
+     */
+    public static <T, R> Publisher<R> map(Publisher<T> source, Function<T, R> map) {
+        return subscriber -> {
+            source.subscribe(new Subscriber<T>() {
+                @Override
+                public void onSubscribe(Subscription s) {
+                    subscriber.onSubscribe(s);
+                }
+
+                @Override
+                public void onNext(T t) {
+                    try {
+                        R r = map.apply(t);
+                        subscriber.onNext(r);
+                    } catch (Exception e) {
+                        onError(e);
+                    }
+                }
+
+                @Override
+                public void onError(Throwable t) {
+                    subscriber.onError(t);
+                }
+
+                @Override
+                public void onComplete() {
+                    subscriber.onComplete();
+                }
+            });
+        };
+    }
+
+    /**
+     * Adds a timeout for the first emission from the {@code source}. If the source emits multiple items then this
+     * timeout does <em>not</em> apply for further emissions.
+     *
+     * @param source Source to apply the timeout on.
+     * @param timeoutSignal Source after termination of which, {@code source} will be cancelled, if not already done.
+     *
+     * @param <T> Type of items emitted by {@code source}.
+     *
+     * @return A new {@code Publisher} with timeout applied.
+     */
+    public static <T> Publisher<T> timeout(Publisher<T> source, Publisher<Void> timeoutSignal) {
```

Comment:
Agreed it will be easier to use. However, my intention is to keep the API surface area smaller and hence I did not add overrides. Instead I provided a `timer()` method which in connection with this can be used as 
`timeout(source, timer(executor, interval, unit))`.

Having said that, I don't think we should provide ways to specify timeouts/retries on public APIs. Any practical application can not use ReactiveSocket without using a function composition library. When they use one, they can use that to provide timeouts and retries. We should just make sure that we have cold sources with no side-effects.

That is my opinion, but I am open to adding more methods here when convinced 😄